Shared/GhostManager.gd

```diff
@@ -0,0 +1,196 @@
+extends Node
+class_name GhostManager
+
+signal ghost_hit(victim_player_id, victim_timeline_index, perpetrator_player_id, perpetrator_timeline_index)
+signal quiet_ghost_hit(victim_player_id, victim_timeline_index, perpetrator_player_id, perpetrator_timeline_index)
+signal new_record_data_applied(player)
+
+onready var _server = get_node("/root/Server")
+onready var _max_ghosts = Constants.get_value("ghosts", "max_amount")
+
+var _ghost_scene = preload("res://Shared/Characters/GhostBase.tscn")
+
+var _game_manager
+var _round_manager
+var _action_manager
+var _character_manager
+
+# Timeline index <-> ghost
+# holds every ghost 
+var _ghosts: Array = []
+# holds ghosts split by team_id
+var _seperated_ghosts: Array = [[],[]]
+
+var _game_phase_start_time = -1
+# previous deaths that could be relevant for current round
+var _previous_ghost_deaths: Array = []
+# deaths we recorded during the current round (will get added to above array after the round)
+var _new_previous_ghost_death: Array = []
+# previous deaths are ordered by time so we only ever have to check one index 
+# and only after it's time has been reached, the following ones could be relevant
+var _current_ghost_death_index = 0
+
+func _ready():
+	set_physics_process(false)
+
+func _physics_process(delta):
+	_look_for_previous_death()
+	for ghost in _ghosts:
+		ghost.update(delta)
+
+func init(game_manager,round_manager,action_manager, character_manager):
+	_game_manager = game_manager
+	_round_manager = round_manager
+	_action_manager = action_manager
+	_character_manager = character_manager
+	set_physics_process(true)
+	_spawn_all_ghosts()
+	_disable_all_ghosts()
+	_move_ghosts_to_spawn()
+
+func on_preparation_phase_started() -> void:
```

Comment:
Yeah well this is because connections currently are all over the place and are handled differently on server and client, I thought I leave it like that and if we some time implement a mediator class then I'll tidy it up.